### PR TITLE
Functional: improve typing in embedded

### PR DESCRIPTION
**Pull request by havogt**
Improves typing in embedded by encoding semantic meaning in the types/aliases

Additional: remove mypy exclude from iterator_tests.

### CODE REVIEW

File: src/functional/iterator/embedded.py [hunk 1]
```diff
@@ -605,14 +626,37 @@ def getter(indices):
...
+    class ConstantField:
+        def __init__(self, value, dtype):
+            self.value = value
+            self.dtype = dtype
+
+        def __getitem__(self, _: FieldIndexOrIndices):  # TODO return type
+            return value
+
+        @property
+        def axes(self) -> tuple[()]:
```
Comment (egparedes) [inline src/functional/iterator/embedded.py:656]: Does mypy really understand this return type? It makes sense but I had never thought about it before.

File: src/functional/iterator/embedded.py [hunk 2]
```diff
@@ -783,7 +827,7 @@ def impl(*iters: ItIterator):
     return impl
 
 
-def fendef_embedded(fun, *args, **kwargs):  # noqa: 536
+def fendef_embedded(fun, *args, **kwargs):
```
Comment (egparedes) [inline src/functional/iterator/embedded.py:830]: Missing type annotations

**Review by egparedes — COMMENTED** [review]
It looks good. Only a few minor comments, fixes and suggestions.

File: src/functional/ffront/decorator.py [hunk 3]
```diff
@@ -244,7 +244,7 @@ def __call__(self, *args, offset_provider: dict[str, Dimension], **kwargs) -> No
                 )
             if not isinstance(param.type, ct.FieldType):
                 continue
-            if args[param_idx].array is None:
+            if not hasattr(args[param_idx], "array"):
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/functional/ffront/decorator.py:247]: It's changing behavior and it's actually wrong, it should be

```suggestion
            if not hasattr(args[param_idx], "__array__"):
```

Previously all fields were implemented in terms of `LocatedFieldImpl`, especially `index_field` and `constant_field`. Therefore `array` was always defined, but set to None for fields that don't have it. It was not my intention to require from all fields to have `array`.
I'd like to discuss this point with you, to improve LocatedFieldImpl.

File: src/functional/iterator/embedded.py (hunk 1 above)
Comment (havogt) [reply src/functional/iterator/embedded.py:656]: I googled that...

File: src/functional/iterator/embedded.py (hunk 2 above)
Comment (havogt) [reply src/functional/iterator/embedded.py:830]: If you like to help me make it fancy, we can try to match `*args` of `fendef_embedded` to `*args` of fun...
For now I do `fun: Callable[..., None], *args: Any, **kwargs: Any`

File: src/functional/iterator/embedded.py [hunk 4]
```diff
@@ -459,19 +487,16 @@ def deref(self) -> Any:
...
-    offset_provider: dict[str, Any],
+    offset_provider: OffsetProvider,
     *,
-    column_axis: Optional[str],
+    column_axis: Optional[Tag],
 ) -> MDIterator:
     # TODO(havogt): support nested tuples
     axes = inp[0].axes if isinstance(inp, tuple) else inp.axes
-    sparse_dimensions: list[str] = []
+    sparse_dimensions = list[Tag]()
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/functional/iterator/embedded.py:499]: Is your suggestion generally preferred?

File: src/functional/iterator/embedded.py [hunk 5]
```diff
@@ -251,24 +270,37 @@ def less(first, second):
...
+    pos: Position, tag: Tag, index: IntIndex, *, offset_provider: OffsetProvider
 ) -> MaybePosition:
     assert pos is not None
     if tag in pos and pos[tag] is None:  # sparse field with offset as neighbor dimension
         new_pos = pos | {tag: index}
         return new_pos
+
     assert tag in offset_provider
     offset_implementation = offset_provider[tag]
     if isinstance(offset_implementation, Dimension):
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/functional/iterator/embedded.py:306]: I don't think so, let's discuss.

File: src/functional/iterator/embedded.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/functional/iterator/embedded.py:499]: ah, it's a bug. Can you tell me what I did?

**Review by egparedes — APPROVED** [review]
It looks good.

**Pull request merged**